test/unit/DominantJuice.unit.t.sol

```diff
@@ -0,0 +1,962 @@
+        vm.prank(rando);
+        dominantJuice.payParams(payParamsData);
+    }
+
+    // PASSING - (unimplemented feature error)
+    function test_payParams_happyPathReturnsMemoryVariables() public bonusDeposited {
+        payParamsData.payer = pledger1;
+        payParamsData.weight = 1;
+        payParamsData.memo = "juice";
+        payParamsData.amount = JBTokenAmount(ethToken, MIN_PLEDGE_AMOUNT, 18, 1);
+
+        // Error: Unimplemented feature (/solidity/libsolidity/codegen/ArrayUtils.cpp:228):
+        // Copying of type struct JBPayDelegateAllocation3_1_1 memory[] memory to storage not yet supported.
+        // delegateAllocations = new JBPayDelegateAllocation3_1_1[](1);
+        // delegateAllocations[0] = JBPayDelegateAllocation3_1_1(dominantJuice, 0, "");
+
+        (uint256 _weight, string memory _memo,) = dominantJuice.payParams(payParamsData);
+        assertEq(payParamsData.weight, _weight);
+        assertEq(payParamsData.memo, _memo);
+        //assertEq(0, delegateAllocations[0].amount);
+
+        // Assert that there are no major contract changes when payParams is called directly:
+        DominantJuice.FundingStatus memory status = dominantJuice.getCycleFundingStatus();
+
+        assertEq(0, status.totalPledged);
+        assertEq(0, status.percentOfGoal);
+        assertFalse(status.isTargetMet);
+        assertFalse(status.hasCycleExpired);
+        assertEq(TOTAL_REFUND_BONUS, dominantJuice.getBalance());
+        (uint256 pledgerW, uint256 totalW) = dominantJuice.getPledgerAndTotalWeights(pledger1);
+        assertEq(0, pledgerW);
+        assertEq(0, totalW);
+    }
+
+    // payParams() function is non-payable, the parent function that calls it is nonReentrant, and there are only if
+    // statements and logic to satisfy Juicebox architecture, thus not many unit tests here. This function can be
+    // called directly, but nothing would happen. The only way to pledge correctly is by calling
+    // `JBPayoutRedemptionPaymentTerminal3_1_1.pay()`.
+
+    ///////////////////////
+    // didPay() Tests
+    ///////////////////////
+
+    // PASSING
+    function test_didPay_revertsIfPaymentSent(uint256 _amount) public {
+        _amount = bound(_amount, 1, 2000 ether);
+        didPayData.projectId = projectID;
+
+        vm.expectRevert("Pledges should be made through JB website.");
+        vm.prank(rando);
+        dominantJuice.didPay{value: _amount}(didPayData);
+    }
+
+    // PASSING - call reverts when terminal calls with empty JBDidPayData struct (e.g. no project ID)
+    function test_didPay_revertsWithEmptyDataStruct() public {
+        vm.mockCall(
+            address(directory),
+            abi.encodeCall(directory.isTerminalOf, (projectID, ethPaymentTerminal)),
+            abi.encode(true)
+        );
+
+        vm.expectRevert(DominantJuice.IncorrectProjectID.selector);
+        vm.prank(address(ethPaymentTerminal));
+        dominantJuice.didPay(didPayData);
```

Comment:
I’ve run tests and whatever is stored before `setup()` goes back to it’s state as shown. So if you `console.log(didPayData.payer)` in any random test body, it will log out as `0x0000000...`, since it wasn’t populated with anything in `setup()`. It’s my understanding that aside from whatever is done in `setup()`, that all state after a test (function that starts with word `test`) gets wiped and goes back to what it was before the test was run. But regardless, I can def delete the structs and make them memory. I don’t even think I’m using most of them as of right now. I think I just did that to mostly get my bearings and lay out what I would possibly need.